cpp/src/IceBridge/IceBridge.cpp

```diff
@@ -13,124 +13,49 @@ using namespace Ice;
 
 namespace
 {
-
-//
-// Represents a pending invocation.
-//
-class Invocation : public IceUtil::Shared
-{
-public:
-
-    Invocation(const AMD_Object_ice_invokePtr& cb) :
-        _cb(cb)
-    {
-    }
-
-    void success(bool ok, const pair<const Byte*, const Byte*>& results)
-    {
-        _cb->ice_response(ok, results);
-    }
-
-    void exception(const Exception& ex)
-    {
-        _cb->ice_exception(ex);
-    }
-
-    void sent(bool)
-    {
-        _cb->ice_response(true, vector<Byte>()); // For oneway invocations
-    }
-
-private:
-
-    const AMD_Object_ice_invokePtr _cb;
-};
-typedef IceUtil::Handle<Invocation> InvocationPtr;
-
 //
-// Holds information about an incoming invocation that's been queued until an outgoing connection has
+// Holds information about an incoming dispatch that's been queued until an outgoing connection has
 // been established.
 //
-struct QueuedInvocation : public IceUtil::Shared
+struct QueuedDispatch final
 {
     //
-    // The pointers in paramData refer to the Ice marshaling buffer and won't remain valid after
-    // ice_invoke_async completes, so we have to make a copy of the parameter data.
+    // The pointers in p refer to the Ice marshaling buffer and won't remain valid after
+    // ice_invokeAsync completes, so we have to make a copy of the in parameters
     //
-    QueuedInvocation(const AMD_Object_ice_invokePtr& c, const pair<const Byte*, const Byte*>& p, const Current& curr) :
-        cb(c), paramData(p.first, p.second), current(curr)
+    QueuedDispatch(pair<const Byte*, const Byte*> p,
+                   function<void(bool, const pair<const Byte*, const Byte*>&)>&& r,
+                   function<void(exception_ptr)>&& e,
```

Comment:
Why don't we pass these by value?

Reply:
Pass by value would work too; this would just require an extra move in the callee.
Here the && parameters work well because the caller and callee are tightly coupled, and the callee knows it's adopting the caller's function parameters.